### ISSUE TITLE
Chain static template parameter out of memory

### ISSUE THREAD

**Issue by AdeAttwood** [vimeo/psalm#6382]
We are getting an out of memory error when we are chaining our `where()`
functions. This I think is coming from `&static` that is getting joined together
when ever we call another `where()`.

With one `where()` we get a type of

```
ActiveQuery<PostSearch&ActiveQuery<PostSearch&static>>&static
```
And then with two

```
ActiveQuery<PostSearch&ActiveQuery<PostSearch&ActiveQuery<PostSearch&ActiveQuery<PostSearch&static>>&static>&ActiveQuery<PostSearch&ActiveQuery<PostSearch&static>>&static>&static
```

I have replicated the error on `psalm.dev` with no errors. If you then add
another `where()` on line 34 you get an out of memory error.

https://psalm.dev/r/311948d849

```php
$query = self::find()
    ->where(['web_ref' => null])
    ->where(['web_ref' => null])
    ->where(['web_ref' => null]);
```

```php
$query = self::find()
    ->where(['web_ref' => null])
    ->where(['web_ref' => null])
    ->where(['web_ref' => null])
    ->where(['web_ref' => null]);
```

**Comment by AdeAttwood** [comment]
@orklah if there is anything I can do to help fix the issue let me know. I have not worked on the internals of psalm before, but If you point me in the right direction, I will have a go.

**Comment by orklah** [comment]
@AdeAttwood I'm not quite sure where to look, but one thing that we already noticed on this is that the `/** @return static */` in phpdoc does not react the same way as `: static` in signature:
https://psalm.dev/r/ce437b0e65

As a workaround, you could declare `: static` if you have control over the ActiveQuery class or create a stub for it.

As for fixing Psalm, it would require investigating why the phpdoc is different than the signature. Maybe try working your way from `ExistingAtomicMethodCallAnalyzer::analyze:208` I think it's the starting point when Psalm calculate the return type for a specific MethodCall.

**Comment by weirdan** [comment]
It appears `static` is being replaced with current object type which also includes several `static`s, so this leads to `static` being multiplied on each step in geometric progression: https://psalm.dev/r/588038dea3

**Comment by AdeAttwood** [comment]
The issue is that the `static` is getting pared differently if you define it in a docblock than when you define it as a return type, as you said in your previous comments. According to psalm they are different types.

I don't  know if we need to forget that fact, there is a memory issue for the moment, I think that is obscuring the issue. I will get what I have into a WIP PR so you can see what I have been hacking on :+1:

**Comment by orklah** [comment]
Honestly, the PHPDoc way seem to work best so if we can adapt this behaviour to the signature, and it works, I'd go for it.

The memory issue is just the effect of this. We don't need to investigate too much on that if we can fix the source. If we can't, we'll try to limit the impact, but this will hinder analysis

**Comment by AdeAttwood** [comment]
PR is up, I did have one test failing that I found.

```
Psalm\Tests\MagicMethodAnnotationTest::testValidCode with data set "allowMagicMethodStatic"
```

The main goal of this it to get the `FunctionLikeDocblockScanner.php` and `TypeHintResolver.php` working consistently, I now stuck on the test that is passed trying to figure out why the [test is filing](https://github.com/vimeo/psalm/blob/master/tests/MagicMethodAnnotationTest.php#L432-L446) both `$c` and `$d` are of type `C` I can't find the code that determines that.

Let me know what you think.
